### PR TITLE
Better error bars for samples with small contamination in CalculateContamination

### PR DESCRIPTION
**Pull request by davidbenjamin**
@fleharty this is for you.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java [hunk 1]
```diff
@@ -160,15 +162,49 @@ public ContaminationModel(List<PileupSummary> sites) {
                 .mapToDouble(ps -> ps.getTotalCount() * oppositeAlleleFrequency.applyAsDouble(ps))
                 .sum();
 
-        final double contamination = contaminationOppositeDepth / totalDepthWeightedByOppositeFrequency;
+        final double contaminationEstimate = contaminationOppositeDepth / totalDepthWeightedByOppositeFrequency;
+
+        final double coeff1 = homs.stream().mapToDouble(ps -> oppositeAlleleFrequency.applyAsDouble(ps) * ps.getTotalCount()).sum();
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java:167]: Could you specify what equation this is constructing by putting the tex in javadoc?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java [hunk 2]
```diff
@@ -160,15 +162,49 @@ public ContaminationModel(List<PileupSummary> sites) {
...
+        final double coeff2 = homs.stream().mapToDouble(ps ->
+                oppositeAlleleFrequency.applyAsDouble(ps)*(1 - oppositeAlleleFrequency.applyAsDouble(ps)) * MathUtils.square(ps.getTotalCount())
+        ).sum();
+
+        final DoubleUnaryOperator errorFunc = c -> homs.isEmpty() ? 1 : Math.sqrt(coeff1*c*(1-c) + coeff2*c*c) / totalDepthWeightedByOppositeFrequency;
+
+        // we're going to binary search to find the largest contamination whose expected standard error brings it within range of
+        // our estimate.  That is, suppose we estimate a contamination of 0.03 and the standard error of 0.05 is 0.02.  Then 0.05 is
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java:175]: standard error of 0.05

Could you make this clearer?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java [hunk 3]
```diff
@@ -160,15 +162,49 @@ public ContaminationModel(List<PileupSummary> sites) {
...
+        // the upper end of our 1-sigma confidence interval.
+        // this binary search is far from optimized (in fact we could solve this explicitly with a messy closed-form solution)
+        // but it converges to a precision of 1e-6 in 20 iterations of the square root of a linear function.  This is fast enough.
+        double top = 1.0;
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java:179]: Could you either find a standard binary search, or generalize this in MathUtils?

If you add it in MathUtils, create a simple test for it.

**Review by fleharty — COMMENTED** [review]
Very nice change, almost done, just a small ask on binary search.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java (hunk 1 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java:167]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java (hunk 2 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java:175]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java (hunk 3 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java]: done

**Comment by davidbenjamin** [comment]
Back to you @fleharty.

**Comment by davidbenjamin** [comment]
@fleharty github wants your official approving review to merge.

**Review by fleharty — APPROVED** [review]

**Comment by fleharty** [comment]
@davidbenjamin Sorry about that, should work now.

**Pull request merged**